Brightness conversion for Abode dimmers

shred86 opened the pull request:
With AbodePy 0.12.3, dimmers will now work but a conversion of the brightness is required. Additionally, when a brightness value of 100 is sent to Abode, 99 is returned causing AbodePy to throw an error so this component will send 99 instead of 100.

## Description:


**Related issue (if applicable):** N/A

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** N/A

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable #13709 

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

shred86 committed:
Brightness conversion for Abode dimmers

With AbodePy 0.12.3, dimmers will now work but a conversion of the brightness is required. Additionally, when a brightness value of 100 is sent to Abode, 99 is returned causing AbodePy to throw an error so this component will send 99 instead of 100.

homeassistant commented:
Hi @shred86, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#13711).

Once you do that we will be able to review and accept this pull request.

Thanks!

syssi commented on homeassistant/components/light/abode.py:
```diff
@@ -51,7 +51,12 @@ def turn_on(self, **kwargs):
         if ATTR_BRIGHTNESS in kwargs and self._device.is_dimmable:
-            self._device.set_level(kwargs[ATTR_BRIGHTNESS])
+            # Convert HASS brightness (0-255) to Abode brightness (0-100)
+            brightness = int(kwargs[ATTR_BRIGHTNESS] * 100 / 255)
```
If 99 is the upper limit the range should be mapped to 0...99.

shred86 replied:
It's possible to pass a value of 100 to Abode but for some reason when in the response it returns 99, so AbodePy throws an error when the passed value != return value. I'm new to Python but would something like this be a better solution?

`brightness = int(max(min(kwargs[ATTR_BRIGHTNESS] * 100 / 255, 99), 0))`

syssi replied:
It's just math: `kwargs[ATTR_BRIGHTNESS]` provides a value between 0 and 255:

```
0 * 100 / 255 = 0
128 * 100 / 255 = int(50.19) = 50
255 * 100 / 255 = 100

vs. 

0 * 99 / 255 = 0
128 * 99 / 255 = int(49.69) = 49
255 * 99 / 255 = 99
```

syssi replied:
Checkout this one: https://github.com/home-assistant/home-assistant/pull/12203

It's not exactly your case but provides some support may be.

shred86 replied:
Lol, very good point. ;)

syssi replied:
Could you check another edge case? Is the light turned off if you set a brightness of 0 or is 0 the lowest brightness?

shred86 replied:
When I set the brightness all the way to 0, the light is turned off.

syssi replied:
Perfect!

syssi commented on homeassistant/components/light/abode.py:
```diff
@@ -68,7 +73,8 @@ def is_on(self):
         if self._device.is_dimmable and self._device.has_brightness:
-            return self._device.brightness
+            # Convert Abode brightness (0-100) to HASS brightness (0-255)
+            return int(self._device.brightness) * 255 / 100
```
Does the device return 100% ever here? I assume the upper limit should be 99 again.

shred86 replied:
The max value I see returned is 99 from Abode. Similar to my comment above, would something like this be better suited?

````
brightness = int(max(min(self._device.brightness, 99), 0) * 255 / 100)
return brightness
````

shred86 committed:
Update abode.py

Better implementation of the brightness conversion as it's less code and also provides a better UI experience as very small brightness values (i.e. 1, 2) won't be converted to 0.

shred86 committed:
Update abode.py

Minor changes to pass checks and update to comments

syssi commented on homeassistant/components/light/abode.py:
```diff
@@ -68,7 +70,9 @@ def is_on(self):
-            return self._device.brightness
+            # Convert Abode brightness (0-100) to HASS brightness (0-255)
+            # Abode will return 100 during initialization of Abode
+            return ceil(int(self._device.brightness) * 255 / 100)
```
`99 * 255 / 100 = 252.45`. It should be 255! What about `brightness * 255 / 99`

syssi replied:
Just realized your comment. Hmmm...

shred86 replied:
Yeah I had it set to divide by 99 but noticed after restarting HA, Abode was returning a brightness value of 100 from one of my dimmers that was on which caused it to convert to 257. When you set the brightness to 100 is when it will only return 99.

syssi replied:
In this case I would remap the brightness of 100:

```
brightness = int(self._device.brightness)
if brightness == 100:
  return 255
else
  return ceil(brightness * 255 / 99)
```

shred86 committed:
Updates conversion of brightness

Keeps the brightness value sent and returned from the device response consistent. However, during initialization and when a device refresh is received, Abode can return 100 thus we'll convert that case back to 99.

shred86 commented:
@syssi - Thanks for the help. Just made another update to the brightness conversion and tested on my local HA install. Everything is working as expected.

If this is good to go, recommend we add it with #13709 which has already been approved. Otherwise, brightness functionality is not going to work properly without this conversion for Abode dimmers.

syssi commented:
Could you fix this lint issue: R: 75,12: Unnecessary "else" after "return" (no-else-return)

syssi commented on homeassistant/components/light/abode.py:
```diff
@@ -51,7 +51,9 @@ def turn_on(self, **kwargs):
-            self._device.set_level(kwargs[ATTR_BRIGHTNESS])
+            # Convert HASS brightness (0-255) to Abode brightness (0-99)
+            # If 100 is sent to Abode, response is 99 causing an error
+            self._device.set_level(ceil(kwargs[ATTR_BRIGHTNESS] * 99 / 255))
```
Please device by 255.0:

```
>>> ceil(1 * 99 / 255)
0.0
>>> ceil(1 * 99 / 255.0)
1.0
```

syssi commented on homeassistant/components/light/abode.py:
```diff
@@ -68,7 +70,13 @@ def is_on(self):
+                return 255
+            else:
+                # Convert Abode brightness (0-99) to HASS brightness (0-255)
+                return ceil(brightness * 255 / 99)
```
Same here

shred86 committed:
Minor updates to conversion
